### ISSUE TITLE
Version the Store

### ISSUE THREAD

**Issue by larskarlitski** [osbuild/osbuild-composer#349]
@ondrejbudai:
> We want to start maintaining backwards compatibility soon, so we need to version the store to be able to do migrations (that's the word!) from an older version to a newer one.

I think adding a simple `version: 1` field now buys us all the flexibility in the future. Is there anything I'm not thinking about?

**Comment by msehnout** [comment]
Maybe some legacy from lorax? Are there any tools that could depend on this?

**Comment by larskarlitski** [comment]
> Maybe some legacy from lorax? Are there any tools that could depend on this?

Since the store is in a directory private to the service (`/var/lib/osbuild-composer`), I don't think any tools can depend on it. I think it's only about different versions of osbuild-composer accessing the store at different points in time.

Lorax (correctly :wink:) stores its state in a different format under `/var/lib/lorax`.

**Comment by teg** [comment]
Another thought: why don't we just name the files srore-1.json, store-2.json etc? Then we could have an external migration tool if we won't, and if we don't, then worst case you start with a blank state (and the old one is untouched).

**Comment by larskarlitski** [comment]
> Another thought: why don't we just name the files srore-1.json, store-2.json etc? Then we could have an external migration tool if we won't, and if we don't, then worst case you start with a blank state (and the old one is untouched).

Genius. I really like this.

**Comment by bcl** [comment]
Ohh, nice! I like that, it makes sure that the migration cannot trash the old copy.

**Comment by bcl** [comment]
Any more thoughts on this? I think I should tackle this before doing #592 and #593 which are going to require changes to the store. I like the idea of a new store filename and will go with that if there are no objections.

Questions I have are
* Do we need to support moving backwards? I don't think so, if you need to use the old store file run a previous version of osbuild-composer.
* How do we alert the user that a new version of the store is available? /api/status could be used for that.
* How do we document the migration process and expose that to the user when their data suddenly vanishes?
* Should migration be handled by a CLI tool, an API call, both?
* How do we handle a situation like: upgrade, empty new store, users builds some things, then wants to migrate? Overwrite changes? Only migrate non-conflicting items (eg. sources and blueprint names could exist in both places)? I think I prefer only migrating things that don't conflict, assume if they added it to the new store they want it left alone.

**Comment by ondrejbudai** [comment]
Can we make the migrations just an internal thing, not visible to the user? Upgrade the store to the newest version when starting composer. Letting the user to choose when to upgrade means that we have to support multiple stores in our codebase, which sounds quite messy to me.

**Comment by bcl** [comment]
In previous discussions with @teg I think he preferred a manual migration. But at the same time I don't think we can support multiple versions, I addressed this in the first point -- run an old version if you want to use the old data.

From a user perspective the cleanest thing to do is detect a change and migrate at startup. We can still use versioned store files for that which would protect the old data.

**Comment by teg** [comment]
>     * Do we need to support moving backwards? I don't think so, if you need to use the old store file run a previous version of osbuild-composer.

I agree, we should not support this. Would be a nightmare to test, and of questionable value.

>     * How do we alert the user that a new version of the store is available? /api/status could be used for that.

I would have thought we don't need to alert them, but that store conversion would happen as part of upgrade on start-up or something like that? (I would have preferred to not have it in-process, but still somehow automatic).

>     * How do we document the migration process and expose that to the user when their data suddenly vanishes?

I would say that we just do this automatically, so no real documentation necessary. It is all internal state that the user should not know about.

>     * Should migration be handled by a CLI tool, an API call, both?

I liked your idea of doing a tool that migrates from version N to N+1, and then just keep each of the tools around and run them in sequence to get up to speed. A CLI tool in other words. Not sure how to trigger them, but maybe just in an ExecStartPre of the systemd unit? Or a post-install script? I would make this as simple as possible, and that each version just exits without doing anything if either store version N does not exist, or version N+1 already does.

This does mean that if something goes pearshaped, and either we forget to ship the upgrade tool or it somehow does not get run before store version N+1 is created, the old state will appear lost (but it will still be there of course, and possible to resuscitate manually). I would not put much effort into making such a rescue operation discoverable/easy/documented, just possible.

>     * How do we handle a situation like: upgrade, empty new store, users builds some things, then wants to migrate? Overwrite changes? Only migrate non-conflicting items (eg. sources and blueprint names could exist in both places)? I think I prefer only migrating things that don't conflict, assume if they added it to the new store they want it left alone.

If this somehow happens I would say we ignore the old state (though leave it in place). This is not a supported workflow, but in case of emergency, it would be possible to get back to the old state somehow.

**Comment by bcl** [comment]
Ah! So an automatic process using tools that can also be used from the cmdline in case of some unforseen failure. I'm not sure if there is any benefit to making these separate programs, you'll still need one program to figure out which ones to run so it makes more sense to me to just make it one thing, but with clearly separate functions for migrating between versions.

**Comment by teg** [comment]
> Ah! So an automatic process using tools that can also be used from the cmdline in case of some unforseen failure. I'm not sure if there is any benefit to making these separate programs, you'll still need one program to figure out which ones to run so it makes more sense to me to just make it one thing, but with clearly separate functions for migrating between versions.

Sorry for the late reply. I think you are right. We should just do this internally, but split the code for each version into separate files. I now added `json.go` to the store package which uses `fooV0` structs. We could introduce `json-1.go` with `fooV1` structs in the future, and then implement a translation from `V0` to `V1`. Is that what you had in mind?

**Comment by bcl** [comment]
Yes, that's what I was thinking. Haven't had a chance to actually write anything yet, but it's on my list.
